README.md
# samplerr

## Introduction

The main goal of this project is to provide a means for long term relevant storage for your metrics.
It borrows some of [rrdtool](http://rrdtool.org/)'s concepts and leverages the power of a modern storage backend: [elasticsearch](http://elastic.co/products/elasticsearch).

The idea is to downsample metrics to multiple sampling rates by consolidating those using meaningful aggregation functions: multiple archive stores with different resolutions.
Different resolution archives are mainly useful for two reasons:

1. Keep storage space in bounds
2. Keep data amount in bounds at query time

Different consolidation functions (*e.g.* min, max, avg, *etc.*) are mainly useful for keeping track of what matters in the metrics you keep.

samplerr keeps storage low and client queries fast by purging high-resolution data periodically and creates
[elasticsearch aliases](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html) to point the clients to the highest available resolution.

![elasticsearch aliases](doc/samplerr.gif)

## How it works

![sampler diagram](doc/samplerr.png)

In this example, samplerr ingests a metric which has a 5s interval. It then downsamples it to 3 different archives with different consolidation functions.
It keeps different retention policies for each elasticsearch index. For instance, the highest resolution data (30s) is kept for two days, while the lowest resolution (3h) is kept for 1 year.
The disk footprint is the same for all three data stores.

## Features

* multiple resolution archives
* consolidation functions
* constant round robin storage footprint per metric with respect to time
* transparent query across all archives

Its architecure is modular, so you can use any of its following main functions:

* *Downsample* metrics using consolidation functions
* *Persist* metrics to the storage backend
* *Rotate* archive references
* *Purge* expired archives

## Implementation

The current implementation:

* is a [riemann](http://riemann.io/) plugin
* writes your metrics to [elasticsearch](http://elastic.co/products/elasticsearch)
* aggregates data using arbitrary clojure functions
* aggregates data in realtime into different round robin time-based elasticsearch indices (archives)
* manages your time based elasticsearch aliases to point to highest possible resolution data
* ensures your metric stays within storage boundaries

## Installation

After cloning the repo, you can build the plugin using [leiningen](/technomancy/leiningen)

```
lein uberjar
```

This will create a plugin jar named `samplerr-x.y.z-SNAPSHOT-standalone.jar` which you can include into your *java classpath*, *e.g.*:

```
java -cp /usr/lib/riemann/riemann.jar:/usr/lib/riemann/samplerr-0.1.1-SNAPSHOT-standalone-up.jar riemann.bin start /etc/riemann/riemann.config
```

On debian or redhat you could also add the classpath using the `EXTRA_CLASSPATH` variable available respectively in `/etc/default/riemann` or `/etc/sysconfig/riemann`.

## Synopsis

```clojure
(load-plugins)
(require '[clj-time.core :as t])

(let [elastic      (samplerr/connect {:hosts ["http://localhost:9200"]})
      index-prefix ".samplerr"
      alias-prefix "samplerr"
      cfunc        [{:func samplerr/average :name "avg"}
                    {:func samplerr/minimum :name "min"}
                    {:func samplerr/maximum :name "max"}]
      archives     [{:tf "YYYY.MM.dd" :step (t/seconds 20) :ttl   (t/days 2) :cfunc cfunc}
                    {:tf "YYYY.MM"    :step (t/minutes 10) :ttl (t/months 2) :cfunc cfunc}
                    {:tf "YYYY"       :step    (t/hours 1) :ttl (t/years 10) :cfunc cfunc}]
      rotate       (samplerr/periodically-rotate {:interval (t/days 1) :conn elastic :index-prefix index-prefix :alias-prefix alias-prefix :archives archives})
      persist      (batch 1000 10 (samplerr/persist {:index-prefix index-prefix :index-type "samplerr" :conn elastic}))]

  (streams
    (where (tagged "collectd")
      (by [:host :service]
       (samplerr/down archives persist))))
  rotate)
```

## Usage

`samplerr` provides five high-level functions, two of which are stream functions.

### Stream functions

#### `(down archives & children)`

This stream function splits streams by archive and consolidation functions.
It conveniently passes on events to child streams, for example to send those to elasticsearch using the `persist` stream function.

The sequence `archives` should contain at least one archive. Each archive describes the aggregation that shall be performed and the target archive:

```clojure
(def archives [{:tf "YYYY.MM.dd" :step   20 :cfunc cfunc}
               {:tf "YYYY.MM"    :step  600 :cfunc cfunc}
               {:tf "YYYY"       :step 3600 :cfunc cfunc}])
```

* `:tf` time format string to be used to target the archive. This will be used by `persist` to target the corresponding elasticsearch index. This will be parsed by `clj-time.format` and must thus be valid. Example: the event `{:time 1458207113000 :metric 42}` will be indexed to elasticsearch into `.samplerr-2016.03.17`, `.samplerr-2016.03` and `.samplerr-2016` concurrently with the above config.
* `:step` contains the consolidation time interval to be used to accumulate events to be aggregated using `cfunc`. This is the equivalent of `rrdtool`'s step, and represents the resolution of your time series.
* `:cfunc` contains the list of consolidation functions to be used.

Consolidation functions are a hash map containing two keys:

```clojure
(def cfunc [{:func samplerr/average :name avg}
            {:func samplerr/minimum :name min}
            {:func samplerr/maximum :name max}])
```

* The value of `:func` contains the stream function to be used for consolidation. It should accept one parameter corresponding to the `:step` interval. **the interface may change in the future**
* The value of `:name` will be used as an attribute to the consolidated events, and subsequently be indexed using elasticsearch. Following up on the above example: the same event stream will be indexed to 9 elasticsearch documents: one per archive and per cfunc. For instance: `{"@timestamp": "2016-03-17T10:31:53+01:00", "metric": 42, "cfunc": "avg", "_index": ".samplerr-2016.03.17"}`

`samplerr` provides some commonly used cfuncs like `average`, `minimum` and `maximum` which are described in the corresponding section.

#### `(persist options & children)`

This stream function sends events processed by `down` to the storage backend (elasticsearch). It is configured using the hash-map `options`:

```clojure
(def options {:index-prefix index-prefix :index-type index-type :conn es-conn-handle})
```

* `:index-prefix` points to the string to be prefixed to the elasticsearch index. The event's time formatted using the archive's `:tf` will be appended to that prefix.
* `:index-type` elasticsearch document type
* `:conn` connection handle to the elasticsearch REST endpoint. This can be a [`qbits.spandex/client` endpoint](https://github.com/mpenet/spandex/blob/master/src/clj/qbits/spandex.clj#L33), or our wrapped one called `connect`

Events should contain the riemann attribute `:tf` which will route them to the appropriate archive.

### Other functions

#### `(connect)`

This is a proxy to `qbits.spandex/client`

#### `(rotate {:conn es-conn-handle :index-prefix index-prefix :alias-prefix alias-prefix :archives archives)`

This will manage elasticsearch aliases.
Aliases will be created for each `archive` by concatenating `index-prefix` with the `:tf` formatted date and will point to the first *unexpired* index (prefix `index-prefix`). Expiry is computed using the archive's `:ttl`.
The idea behind this is that clients will query elasticsearch using the aliases. Most high-level clients (*e.g.* [grafana], [kibana]) can only point to one time-base index pattern, *e.g.* `foo-YYYY.MM.dd`.

`samplerr` will transparently position aliases pointing to the highest possible resolution archive that overlaps with it and that is not expired. The algorithm is roughly the following:

* for each index matching `<index-prefix>*`
  * is the ttl expired?
    * YES: move all its aliases to the next unexpired period
    * NO:
      * find archive it belongs to
      * parse the time of the beginning of its period using `:tf`
      * add an alias `<alias-prefix>-<parsed-time>`

The usual way to use this function is either:

* periodically using `periodically-rotate`
* triggered by an event in the stream. For instance you could trigger the rotation when the day changes

#### `(periodically-rotate {:interval periodicity :conn es-conn-handle :index-prefix index-prefix :alias-prefix alias-prefix :archives archives)`

This function will call `rotate` every `periodicity` time interval. The first argument should be given in terms of a `org.joda.time/PeriodType` object conventiently provided by `clj-time.core` using *e.g.* `hours`, `days`, *etc.*

Note that the first rotation will not take effect immediately after riemann startup.
Also note that configuration reloads will work as expected.

##### Example

Take the example in the [synopsis](#synopsis) section. Let's say today is 2016-02-01 at 03:14 PM and
riemann started exactly 2 days ago. `samplerr/rotate` fires up and processes the elasticsearch indices:

* `.samplerr-2016.02.01` is younger than two days: create alias `samplerr-2016.02.01`
* `.samplerr-2016.01.31` is younger than two days: create alias `samplerr-2016.01.31`
* `.samplerr-2016.01.30` is two days old: expired! move its aliases to `.samplerr-2016.01`
* `.sampler-2015.02` is younger than two months: create alias `sampler-2015.02`
* `.sampler-2015.01` is younger than two months: create alias `sampler-2015.01`
* `.sampler-2014.12` is two months old: expired! move its aliases to `.samplerr-2014`
* …

#### `(purge {:conn es-conn-handle :index-prefix index-prefix :archives archives)`

This function will **DELETE** expired indices. Use with care.

The usual way to use this function is either:

* periodically using `periodically-purge`
* triggered by an event in the stream. For instance you could trigger the purge when the disk space is full on the elasticsearch node

#### `(periodically-purge {:interval periodicity :conn es-conn-handle :index-prefix index-prefix :archives archives)`

This function will call `purge` periodically.

## Development

At the time of writing the contributors of this project are Fabien Wernli and some code from the elasticsearch integration was borrowed from [tnn1t1s](https://github.com/tnn1t1s/riemann-elastic) which itself borrowed from [kiries](https://github.com/threatgrid/kiries).



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- bar

### Fixed
- foo

## 0.6.4 - 2018-09-13
- fix bug that made rotation still throw on ES6.x
- hopefully last embarassing attempt at fixing this bug

## 0.6.3 - 2018-09-12
- fix bug that made rotation still throw on ES6.x

## 0.6.2 - 2018-09-11
- fix bug that made rotation/purge throw on ES6.x

## 0.6.1 - 2018-08-23
- bump spandex dep to 0.6.4
- fix some typos in README (thanks boernd and brutasse)
- improve error handling

## 0.4.1 - 2017-07-11
- move from clojurewerkz/elastisch to qbits/spandex

## 0.3.6 - 2017-06-06
- adds elasticsearch 5.x compatibility
- bump dep versions to match riemann 0.2.13

## 0.3.5 - 2017-01-13

### Fixed
- low frequency (lower than step) events were lost

## 0.3.3 - 2016-09-05

### Fixed
- events were lost e.g. not part of aggregation
  this is now fixed in average/minimum/maximum/counter

## 0.3.2 - 2016-03-31

### Fixed
- purge is now fixed (was noop)

## 0.3.1 - 2016-03-22

### Changed
- remove at-at dependency
- change rotate/purge semantics

## 0.2.1 - 2016-03-17

### Changed
- change semantics of higher-level functions
- implementes alias rotation

## [0.1.1] - 2016-02-05
- initial release



project.clj
(defproject samplerr "0.6.5pre0"
  :description "riemann plugin to aggregate data in a round-robin fashion to elasticsearch"
  :url "http://github.com/samplerr/samplerr"
  :license {:name "EPL-1.0"
            :url "https://spdx.org/licenses/EPL-1.0.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [cheshire "5.7.0"]
                 [clj-time "0.13.0"]
                 [riemann "0.2.13"]
                 [org.clojure/tools.logging "0.3.1"]
                 [cc.qbits/spandex "0.6.4"]]
  :plugins [[lein-rpm "0.0.5"
             :exclusions [org.apache.maven/maven-plugin-api
                          org.codehaus.plexus/plexus-container-default
                          org.codehaus.plexus/plexus-utils
                          org.clojure/clojure
                          classworlds]]
            [org.apache.maven/maven-plugin-api "2.0"]]
            )



src/riemann/plugin/samplerr.clj
; Copyright (c) Centre de Calcul de l'IN2P3 du CNRS
; Contributor(s) : Fabien Wernli (2016)

(ns riemann.plugin.samplerr
  "A riemann plugin to downsample data in a RRDTool fashion into elasticsearch"
  (:use [clojure.tools.logging :only (info error debug warn)]
        [riemann.common :only [member?]]
        [riemann.time :only [unix-time]])
  (:require [cheshire.core :as json]
            [clj-time.format]
            [clj-time.core]
            [clj-time.coerce]
            [clojure.edn :as edn]
            [clojure.java.io :as io]
            [qbits.spandex :as es]
            [riemann.config]
            [riemann.core :as core]
            [riemann.service :as service]
            [riemann.streams :as streams]))

(defn ^{:private true} keys-to-map [key-map] 
  (reduce-kv (fn [key-map k v] 
                 (assoc-in key-map 
                              (clojure.string/split (name k) #"\.")
                              (if (map? v) 
                                  (keys-to-map v) 
                                  v))) 
                                {} key-map))


(def ^{:private true} format-iso8601
  (clj-time.format/with-zone (clj-time.format/formatters :date-time-no-ms)
    clj-time.core/utc))

(defn ^{:private true} iso8601 [event-s]
  (clj-time.format/unparse format-iso8601
                           (clj-time.coerce/from-long (* 1000 event-s))))

(defn ^{:private true} safe-iso8601 [event-s]
  (try (iso8601 event-s)
    (catch Exception e
      (warn "Unable to parse iso8601 input: " event-s)
      (clj-time.format/unparse format-iso8601 (clj-time.core/now)))))

(defn ^{:private true} stashify-timestamp [event]
  (->  (if-not (get event "@timestamp")
         (let [time (:time event)]
           (if (nil? time) (info event))
           (assoc event "@timestamp" (safe-iso8601 (long time))))
         event)
       (dissoc :time)
       (dissoc :ttl)))

(defn ^{:private true} edn-safe-read [v]
  (try
    (edn/read-string v)
    (catch Exception e
      (warn "Unable to read supposed EDN form with value: " v)
      v)))

(defn ^{:private true} message-event [event]
  (keys-to-map
    (into {}
        (for [[k v] event
              :when v]
          (cond
           (= (name k) "_id") [k v]
           (.startsWith (name k) "_")
           [(.substring (name k) 1) (edn-safe-read v)]
           :else
           [k v])))))

(defn ^{:private true} elastic-event [event message]
  (let [e (-> event
              stashify-timestamp)]
    (if message
      (message-event e)
      e)))

(defn ^{:private true} riemann-to-elasticsearch [events message]
  (->> [events]
       flatten
       (remove streams/expired?)
       (map #(elastic-event % message))))

(defn connect
  "Connect to ElasticSearch"
  [& argv]
  (apply es/client argv))

  
(defn- make-index-timestamper [event]
  (let [formatter (clj-time.format/formatter (eval (get event "tf")))]
    (fn [date]
      (clj-time.format/unparse formatter date))))

(defn persist
  "bulk index to ES"
  [{:keys [conn index-type index-prefix message]
               :or {index-type "samplerr" index-prefix ".samplerr"
                    message true}} & children]
    (fn [events]
      (let [esets (group-by (fn [e] (let [index-namer (make-index-timestamper e)]
                              (str index-prefix (index-namer
                               (clj-time.format/parse format-iso8601 
                                                      (get e "@timestamp"))))))
                            (riemann-to-elasticsearch events message))]
        (doseq [es_index (keys esets)]
          (let [raw (get esets es_index)
                bulk-create-items
                (interleave (map #(if-let [id (get % "_id")]
                                    {:index {:_type index-type :_index es_index :_id id}}
                                    {:index {:_type index-type :_index es_index}}
                                    )
                                 raw)
                            raw)]
            (when (seq bulk-create-items)
              (try
                (let [response (es/request conn {:url "_bulk" :body (es/chunks->body bulk-create-items) :content-type "application/x-ndjson" :method :post})
                      res (:body response)
                      by_status (frequencies (map :status (map :index (:items res))))
                      total (count (:items res))
                      succ (filter :_version (map :index (:items res)))
                      failed (filter :error (map :index (:items res)))]
                  (info "elasticized" total " (total) " by_status " docs to " es_index "in " (:took res) "ms")
                  (debug "Failed: " failed))
                (catch Exception e
                  (error "Unable to bulk index:" e)))))))))

(defn ^{:private true} resource-as-json [resource-name]
  (json/parse-string (slurp (io/resource resource-name))))


(defn ^{:private true} file-as-json [file-name]
  (try
    (json/parse-string (slurp file-name))
    (catch Exception e
      (error "Exception while reading JSON file: " file-name)
      (throw e))))


;;;;;;
;;;;;;
;;;;;;

(defn- new-interval?
  [acc event]
  (when-let [acc-time (:time acc)]
    (let [event-time (:time event)
          age (- event-time acc-time)
          step (:step event)]
      (>= age step))))

; cfuncs
(defn sum
  [interval & children]
  (streams/sreduce
    (fn [acc event]
      (if (nil? acc)
        event
        (if (new-interval? acc event)
          (assoc event :parent (dissoc acc :parent))
          (assoc event :time (:time acc) :metric (+ (:metric event) (:metric acc))))))
    nil
    (streams/where (contains? event :parent)
      (streams/smap #((comp (fn [e] (dissoc e :parent))
                            :parent) %)
        (apply streams/sdo children)))))


(defn counter
  [interval & children]
  (streams/sreduce
    (fn [acc event]
      (if (nil? acc)
          (assoc event :metric 1)
          (if (new-interval? acc event)
              (assoc event :metric 1 :parent (dissoc acc :parent))
              (assoc event :time (:time acc) :metric (+ 1 (:metric acc))))))
    nil
    (streams/where (contains? event :parent)
      (streams/smap #((comp (fn [event] (dissoc event :parent)) :parent) %)
        (apply streams/sdo children)))))

(defn average
  [interval & children]
  (streams/sreduce
    (fn [acc event]
      (if (nil? acc)
        (assoc event :sum (:metric event) :count 1)
        (if (new-interval? acc event)
          (assoc event :sum (:metric event) :count 1 :parent (dissoc acc :parent))
          (assoc event :time (:time acc) :count (+ 1 (:count acc)) :sum (+ (:metric event) (:sum acc))))))
    nil
    (streams/where (contains? event :parent)
      (streams/smap #((comp (fn [e] (assoc e :metric (/ (:sum e) (:count e))))
                            (fn [e] (dissoc e :parent))
                            :parent) %)
        (apply streams/sdo children)))))

(defn extremum
  [efunc interval children]
  (streams/sreduce
    (fn [acc event]
      (if (new-interval? acc event)
        (assoc event :parent (dissoc acc :parent) :orig-time (:time event))
        (if (efunc (:metric event) (:metric acc))
          (assoc event :time (:time acc) :orig-time (:time event))
          acc)))
    (streams/where (contains? event :parent)
      (streams/smap #((comp (fn [e] (dissoc e :orig-time))
                            (fn [e] (assoc e :time ((some-fn :orig-time :time) e)))
                            (fn [e] (dissoc e :parent))
                            :parent) %)
        (apply streams/sdo children)))))

(defn maximum
  [interval & children]
  (extremum >= interval children))

(defn minimum
  [interval & children]
  (extremum <= interval children))

(defn- to-seconds*
  "takes a clj-time duration and converts it to seconds"
  [dateobj]
  (clj-time.core/in-seconds dateobj))
(def to-seconds (memoize to-seconds*))

(defn- to-millis*
  "takes a clj-time duration and converts it to milliseconds"
  [dateobj]
  (clj-time.core/in-millis dateobj))
(def to-millis (memoize to-millis*))

(defn down-n-cf
  "takes map of archive parameters and sends time-aggregated data to children"
  [{:keys [cfunc step tf] :as args :or {cfunc {:name "avg" :func average}}} & children]
  (let [cfunc_n (:name cfunc)
        cfunc_f (:func cfunc)
        seconds (to-seconds step)]
    (streams/with {:step seconds :cfunc cfunc_n :tf tf}
      (streams/where metric
        (cfunc_f seconds
          (streams/smap #(assoc % :ttl (max (* seconds 3) (or (:ttl %) 0)) :service (str (:service %) "/" cfunc_n "/" seconds))
            (apply streams/sdo children)))))))

(defn down-n
  "takes map of archive parameters and maps to down-n-cf for all cfuncs"
  [{:keys [cfunc] :as args} & children]
    (apply streams/sdo (map #(apply down-n-cf (assoc args :cfunc %) children) cfunc)))

(defn down
  "takes vector of archives and generates (count vector) down-n streams"
  [archives & children]
  (apply streams/sdo (map #(apply down-n % children) archives)))

;;;
;;; foreign commodity functions
;;; that need license checking

;source http://stackoverflow.com/questions/8641305/find-index-of-an-element-matching-a-predicate-in-clojure
(defn- indices [pred coll]
  (keep-indexed #(when (pred %2) %1) coll))

;;;
;;; alias shifting code
;;;

(defn list-indices
  "lists all indices from an elasticsearch cluster having given prefix"
  [elastic prefix]
  (map name (keys (:indices (:body (es/request elastic {:url [(str prefix "*") :_stats :store] :method :get}))))))

(defn- index-exists?
  "returns true if index exists"
  [elastic index]
  (try
    (es/request elastic {:url [index] :method :head})
    (catch Exception e
      (warn e "index exists caught"))))

(defn matches-timeformat?
  "returns true if datestr matches timeformat"
  [datestr timeformat]
  (try (clj-time.format/parse (clj-time.format/formatter timeformat) datestr) (catch IllegalArgumentException ex false)))

; TODO: should return an additional key which is the parsed clj-time object so we don't need to parse time again
(defn get-retention-policy
  "returns the first retention policy that matches datestr or nil
  example: (get-retention-policy \"2016\" [{:tf \"YYYY.MM.DD\" :ttl 86400} {:tf \"YYYY\" :ttl 315567360}])
  will return {:tf \"YYYY\" :ttl 315567360}"
  [datestr retention-policies]
  (first (filter #(matches-timeformat? datestr (:tf %)) retention-policies)))

(defn get-retention-policy-index
  "returns the first retention policy's index that matches datestr or nil
  example: (get-retention-policy \"2016\" [{:tf \"YYYY.MM.DD\" :ttl 86400} {:tf \"YYYY\" :ttl 315567360}])
  will return {:tf \"YYYY\" :ttl 315567360}"
  [datestr retention-policies]
  (first (indices #(matches-timeformat? datestr (:tf %)) retention-policies)))

(defn- parse-datestr
  "parses datestr using retention-policy"
  [datestr retention-policy]
  (let [tf (:tf retention-policy)]
    (clj-time.format/parse (clj-time.format/formatter tf) datestr)))

(defn parse-retention-policy-date
  "parses datestr using index n of retention-policies"
  [datestr retention-policies n]
  (if n
    (let [retention-policy (nth retention-policies n)]
      (parse-datestr datestr retention-policy))
    (warn (str datestr " does not match any retention policy"))))

(defn- format-datestr
  "formats dateobj using retention-policy"
  [dateobj retention-policy]
  (let [tf (:tf retention-policy)]
    (clj-time.format/unparse (clj-time.format/formatter tf) dateobj)))

(defn format-retention-policy-date
  "formats dateobj using index n of retention-polices"
  [dateobj retention-policies n]
  (let [retention-policy (nth retention-policies n)]
    (format-datestr dateobj retention-policy)))

(defn is-expired?
  "returns true if datestr matches an expired retention policy"
  [datestr retention-policies]
  (let [retention-policy (get-retention-policy datestr retention-policies)
        tf (:tf retention-policy)
        ttl (:ttl retention-policy)
        parsed-time (clj-time.format/parse (clj-time.format/formatter tf) datestr)
        now (clj-time.core/now)
        expiration (clj-time.core/minus now ttl)]
    (clj-time.core/before? parsed-time expiration)))

(defn get-aliases
  "returns aliases of index or empty list"
  [elastic index]
  (keys ((comp :aliases (keyword index))(:body (es/request elastic {:url [index :_alias] :method :get})))))

(defn move-aliases
  "moves aliases from src-index to dst-index"
  [elastic src-index dst-index]
  (info "transfer aliases from" src-index "to" dst-index)
  (let [src-aliases (get-aliases elastic src-index)
        src-actions (map #(hash-map :remove (hash-map :index src-index :alias %)) src-aliases)
        dst-actions (map #(hash-map :add    (hash-map :index dst-index :alias %)) src-aliases)]
    (es/request elastic {:url "/_aliases" :method :post :body {:actions (vec (concat src-actions dst-actions))}})))

(defn add-alias
  "adds alias to index"
  [elastic index es-alias]
  (info "add alias" es-alias "->" index)
  (es/request elastic {:url "/_aliases" :method :post :body {:actions [{:add {:index index :alias es-alias}}]}}))

(defn remove-aliases
  "removes all aliases from index"
  [elastic index]
  (info "remove all aliases from" index)
  (let [aliases (get-aliases elastic index)
        actions (map #(hash-map :remove (hash-map :index index :alias %)) aliases)
        actions {:actions actions}]
    (es/request elastic {:url "/_aliases" :method :post :body actions})))

(defn fresh-index-targets
  "returns collection of unexpired datestrings for dateobj"
  [dateobj retention-policies]
  (for [policy retention-policies 
        :let [now (clj-time.core/now) 
              tf (:tf policy)
              ttl (:ttl policy)]
        :when (clj-time.core/before? (clj-time.core/minus now ttl) dateobj)]
    (clj-time.format/unparse (clj-time.format/formatter tf) dateobj)))

(defn shift-alias
  "matches index with its corresponding retention policy. if expired, shifts its existing aliases to next retention policy. else adds alias if necessary"
  [elastic index index-prefix alias-prefix retention-policies]
  (let [datestr (clojure.string/replace index (re-pattern (str "^" index-prefix)) "")
        retention-policy-index (get-retention-policy-index datestr retention-policies)
        dateobj (parse-retention-policy-date datestr retention-policies retention-policy-index)]
    (if retention-policy-index
      (let [next-dates (fresh-index-targets dateobj retention-policies)
            next-date (first next-dates)
            next-index (str index-prefix next-date)]
        (if (is-expired? datestr retention-policies)
          (do
            (if next-date
              (if (index-exists? elastic next-index)
;;;;;;;stacktrace happens here
;;;;;;;at clojure.core$ex_info.invokeStatic(core.clj:4617) at clojure.core$ex_info.invoke(core.clj:4617) at qbits.spandex$response_ex__GT_ex_info.invokeStatic(spandex.clj:218) at qbits.spandex$response_ex__GT_ex_info.invoke(spandex.clj:215) at qbits.spandex$eval6897$fn__6898.invoke(spandex.clj:227) at qbits.spandex$eval6880$fn__6881$G__6871__6886.invoke(spandex.clj:222) at qbits.spandex$default_exception_handler.invokeStatic(spandex.clj:238) at qbits.spandex$default_exception_handler.invoke(spandex.clj:231) at qbits.spandex$request.invokeStatic(spandex.clj:271) at qbits.spandex$request.invoke(spandex.clj:253) at riemann.plugin.samplerr$index_exists_QMARK_.invokeStatic(samplerr.clj:279) at riemann.plugin.samplerr$index_exists_QMARK_.invoke(samplerr.clj:276) at riemann.plugin.samplerr$shift_alias.invokeStatic(samplerr.clj:388) at riemann.plugin.samplerr$shift_alias.invoke(samplerr.clj:375) at riemann.plugin.samplerr$rotate.invokeStatic(samplerr.clj:454) at riemann.plugin.samplerr$rotate.invoke(samplerr.clj:447) at riemann.plugin.samplerr$rotation_service$rot__7675.invoke(samplerr.clj:470) at riemann.service.ThreadService$thread_service_runner__6478$fn__6479.invoke(service.clj:71) at riemann.service.ThreadService$thread_service_runner__6478.invoke(service.clj:70) at clojure.lang.AFn.run(AFn.java:22) at java.lang.Thread.run(Thread.java:748)
                (if (get-aliases elastic index)
                  (move-aliases elastic index next-index)
                  (add-alias elastic next-index (str alias-prefix datestr)))
                (throw (Exception. (str "can't move aliases from " index " to missing " next-index)))))
            (if (get-aliases elastic index)
              (remove-aliases elastic index)))
          (add-alias elastic index (str alias-prefix datestr)))))))

(defn delete-index
  "deletes index"
  [elastic index]
  (info "delete index" index)
  (es/request elastic {:url [index] :method :delete}))

(defn purge-index
  "deletes index if it matches a timeformat in retention-policies and is expired"
  [elastic index index-prefix retention-policies]
  (let [datestr (clojure.string/replace index (re-pattern (str "^" index-prefix)) "")
        retention-policy-index (get-retention-policy-index datestr retention-policies)
        dateobj (parse-retention-policy-date datestr retention-policies retention-policy-index)]
    (if retention-policy-index
      (if (is-expired? datestr retention-policies)
        (delete-index elastic index)))))

(defn purge
  "deletes all indices matching index-prefix and that are expired according to retention-policies"
  [{:keys [conn index-prefix archives]}]
  (debug "purging")
  (loop [indices (list-indices conn (str index-prefix "*"))]
    (let [current-index (first indices)
          remaining-indices (rest indices)]
      (purge-index conn current-index index-prefix archives)
      (if (not (empty? remaining-indices))
        (recur remaining-indices)))))

(defn purge-service
  "returns a service which schedules a task to purge indices"
  [{:keys [interval conn index-prefix archives enabled?]
    :or {interval (clj-time.core/days 3)
         enabled? true}}]
  (let [interval (to-millis interval)]
    (service/thread-service
      ::samplerr-purge [interval conn index-prefix archives enabled?]
      (fn pur [core]
        (Thread/sleep interval)
        (try
          (if enabled?
            (purge {:conn conn :index-prefix index-prefix :archives archives}))
          (catch Exception e
            (warn e "purge service caught")))))))

(defn periodically-purge
  "adds an index purge service to core"
  [& opts]
  (info "registering purge service with" (apply :interval opts) "interval")
  (let [service (apply purge-service opts)]
    (swap! riemann.config/next-core core/conj-service service :force)))

(defn rotate
  "maps shift-alias to all indices from elastic connection matching index-prefix"
  [{:keys [conn index-prefix alias-prefix archives]}]
  (debug "rotating")
  (loop [indices (list-indices conn (str index-prefix "*"))]
    (let [current-index (first indices)
          remaining-indices (rest indices)]
      (shift-alias conn current-index index-prefix alias-prefix archives)
      (if (not (empty? remaining-indices))
        (recur remaining-indices)))))

(defn rotation-service
  "returns a service which schedules a task to rotate aliases"
  [{:keys [interval conn alias-prefix index-prefix archives enabled?]
    :or {interval (clj-time.core/minutes 5)
         enabled? true}}]
  (let [interval (to-millis interval)]
    (service/thread-service
      ::samplerr-rotation [interval conn alias-prefix index-prefix archives enabled?]
      (fn rot [core]
        (Thread/sleep interval)
        (try
          (if enabled?
            (rotate {:conn conn :index-prefix index-prefix :alias-prefix alias-prefix :archives archives}))
          (catch Exception e
            (warn e "rotation service caught")))))))

(defn periodically-rotate
  "adds an alias rotation service to core"
  [& opts]
  (info "registering rotation service with" (apply :interval opts) "interval")
  (let [service (apply rotation-service opts)]
    (swap! riemann.config/next-core core/conj-service service :force)))

